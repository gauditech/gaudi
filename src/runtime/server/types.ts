import { NextFunction, Request, Response } from "express";

export type ServerHttpMethods = "all" | "get" | "post" | "put" | "patch" | "delete";

export type ServerRequestHandler =
  | ((request: Request, response: Response) => void | Promise<void>)
  | ((req: Request, resp: Response, next: NextFunction) => void | Promise<void>);

export type EndpointConfig = {
  path: string;
  method: ServerHttpMethods;
  handlers: ServerRequestHandler[];
};

export type ServerMiddlewareNextFn = (err?: unknown) => void;
