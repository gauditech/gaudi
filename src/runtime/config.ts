import fs from "fs";

import { initializeContext } from "@src/runtime/server/context";
import { createDbConn } from "@src/runtime/server/dbConn";
import { Definition } from "@src/types/definition";

export type RuntimeConfig = {
  /** Runtime server host name */
  host: string;
  /** Runtime server port number */
  port: number;
  /** Path to generated definition.json file. */
  definitionPath: string;
  /** Folder where runtime should output generated files */
  outputFolder: string;
  /** DB connection URL */
  dbConnUrl: string;
  /** DB default schema */
  dbSchema?: string;
};

/** Read runtime config from environment or provide default values. */
export function readConfig(): RuntimeConfig {
  const host = process.env.GAUDI_RUNTIME_SERVER_HOST ?? "127.0.0.1";
  const port =
    process.env.GAUDI_RUNTIME_SERVER_PORT != null
      ? parseInt(process.env.GAUDI_RUNTIME_SERVER_PORT, 10)
      : 3001;
  const definitionPath = process.env.GAUDI_RUNTIME_DEFINITION_PATH || "definition.json";
  const outputFolder = process.env.GAUDI_RUNTIME_OUTPUT_PATH || ".";

  const dbConnUrl = process.env.GAUDI_DATABASE_URL || "";
  const dbSchema = process.env.GAUDI_DATABASE_SCHEMA;

  return { host, port, definitionPath, outputFolder, dbConnUrl, dbSchema };
}

/** Load definition file and return it's content */
export function loadDefinition(definitionPath: string): Definition {
  // --- read input file
  if (!fs.existsSync(definitionPath)) {
    throw new Error(`Definition file not found: "${definitionPath}"`);
  }
  const definitionStr = fs.readFileSync(definitionPath).toString("utf-8");
  return JSON.parse(definitionStr);
}

export function createAppContext(config: RuntimeConfig) {
  initializeContext({
    dbConn: createDbConn(config.dbConnUrl, {
      schema: config.dbSchema,
    }),
    config: config,
  });
}
