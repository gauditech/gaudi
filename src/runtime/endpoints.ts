import express, { Request } from "express";
import knex from "knex";

const app = express();
const port = 3002;

import {
  PathParam,
  buildEndpointPath,
  queryableFromEndpointTargets,
  selectToSelectable,
} from "@src/builder/query";
import { Definition, QueryDef, QueryDefPath, TargetDef } from "@src/types/definition";

function validatePathParam(param: PathParam["params"][number], val: string): PathVariable {
  switch (param.type) {
    case "integer":
      return { name: param.name, type: param.type, value: parseInt(val, 10) };
    case "text":
      return { name: param.name, type: param.type, value: val };
  }
}

type PathVariable =
  | { name: string; type: "text"; value: string }
  | { name: string; type: "integer"; value: number };

export function setupEndpoints(def: Definition) {
  for (const entrypoint of def.entrypoints) {
    for (const endpoint of entrypoint.endpoints) {
      switch (endpoint.kind) {
        case "get": {
          const { path, params } = buildEndpointPath(endpoint);
          app.get(path, (req: Request, res) => {
            // collect path params
            const vars = params.map((p) => validatePathParam(p, req.params[p.name]));

            // get targets
            const q = queryableFromEndpointTargets(
              def,
              endpoint.targets,
              selectToSelectable(endpoint.response),
              "single"
            );
          });
        }
      }
    }
  }
}

/**
 * Here we go again... Returning a query/able, whatever it may be.


from Org.public_repos.issues
filter
      Org.slug is ...
  and public_repos.slug is ...
  and     
*/
