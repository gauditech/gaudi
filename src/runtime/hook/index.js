/** Returns given value. Can be used for testing. */
module.exports.echo = function ({ value }) {
  return value;
};
