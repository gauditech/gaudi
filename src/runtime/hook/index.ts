/** Concatente all given string parts */
export function concat(...parts: string[]): string {
  return parts.join("");
}
