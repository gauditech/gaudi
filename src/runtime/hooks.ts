import { promises as fs } from "fs";
import path from "path";

import { getExecutionRuntime } from "@src/common/refs";
import { getInternalExecutionRuntimeName } from "@src/composer/executionRuntimes";
import { Definition, ExecutionRuntimeDef, HookCodeDef, HookDef } from "@src/types/definition";

export type ExecutionRuntimeClient = {
  runtimeName: string;
  executeHook: <T>(hook: HookCodeDef, args: Record<string, unknown>) => T;
};

const EXECUTION_RUNTIMES: Record<string, ExecutionRuntimeClient> = {};

export async function executeHook<T>(
  def: Definition,
  hook: HookDef,
  args: Record<string, unknown>
): Promise<T> {
  const execRuntime = getExecutionRuntime(def, hook.runtimeName);
  return (await createExecutionRuntime(execRuntime)).executeHook<T>(hook.code, args);
}

async function createExecutionRuntime(
  runtime: ExecutionRuntimeDef
): Promise<ExecutionRuntimeClient> {
  if (EXECUTION_RUNTIMES[runtime.name] == null) {
    // only local gaudi execution runtime is supported at the moment
    EXECUTION_RUNTIMES[runtime.name] = await createLocalExecutionRuntime(runtime);
  }

  return EXECUTION_RUNTIMES[runtime.name];
}

// ---------- Local execution runtime

export type HookModules = Record<string, Record<string, (_: unknown) => unknown>>;

const HOOKS_FILES_PATTERN = /.+\.[tj]s$/;

export async function createLocalExecutionRuntime(
  runtime: ExecutionRuntimeDef
): Promise<ExecutionRuntimeClient> {
  const modules: HookModules = {};

  const sourcePath = resolveSourcePath(runtime.name, runtime.sourcePath);
  await importHooks(sourcePath, modules);

  return {
    runtimeName: runtime.name,
    executeHook(code, args) {
      switch (code.kind) {
        case "inline": {
          // order of args must be consistent
          const argNames = Object.entries(args).map(([name, _value]) => name);
          const argValues = Object.entries(args).map(([_name, value]) => value);
          const hookBody = [
            // strict mode
            // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode
            "'use strict'",
            // hook fn body
            `return ${code.inline}`,
          ].join("\n");

          // dynamically create new function
          // allows access only to global and function's own scope
          // syntax is `new AsyncFunction(arg1, ..., argN, fnBody)`
          // arguments are positional so order in fn definition (here) and in fn call (below) must be the same
          const hookFn = new AsyncFunction(...[...argNames, hookBody]);

          // TODO: cache inline function - by which key, source code?!

          return hookFn(...argValues);
        }
        case "source": {
          const hookFile = modules[code.file];
          if (hookFile == null) {
            throw new Error(`Hook file "${code.file}" not found`);
          }

          const hook = hookFile[code.target];
          if (hook == null) {
            throw new Error(`Hook "${code.target}" in file "${code.file}" not found`);
          }

          return hook(args);
        }
      }
    },
  };
}

/**
 * Resolves absolute source folder path.
 *
 * If sources path is already absolute it is returned as umchanged.
 *
 * Relative paths are resolved with different base path:
 *  - internal exec runtimes are resolved using `__dirname` as base path
 *  - user defined exec runtimes are resolved using current process working folder
 */
function resolveSourcePath(name: string, sourcePath: string): string {
  if (path.isAbsolute(sourcePath)) return sourcePath;

  const internalExecRuntimes = [getInternalExecutionRuntimeName()];
  if (internalExecRuntimes.includes(name)) {
    return path.resolve(__dirname, sourcePath);
  } else {
    return path.resolve(sourcePath);
  }
}

async function importHooks(sourcePath: string, modules: HookModules) {
  console.log("Loading hooks sources from:", path.resolve(sourcePath));

  async function loadHooksFromDir(dir: string) {
    const entities = await fs.readdir(path.join(sourcePath, dir));

    const promises = entities.map(async (entityFilename) => {
      const hookPath = path.join(dir, entityFilename);
      const entity = path.join(sourcePath, hookPath);

      const stats = await fs.lstat(entity);

      if (stats.isDirectory()) {
        await loadHooksFromDir(hookPath);
      } else if (stats.isFile() && HOOKS_FILES_PATTERN.test(entityFilename)) {
        modules[hookPath] = loadFileAsModule(entity);
      }
    });

    await Promise.all(promises);
  }

  await loadHooksFromDir("");
}

function loadFileAsModule(filepath: string) {
  const absolute = path.resolve(filepath);
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  return require(absolute);
}

// we cannot dynamically create async function using `new Function()` so we'll hijack real async function's prototype to create a new one dynamically
const AsyncFunction = Object.getPrototypeOf(async function () {
  // empty body
}).constructor;
