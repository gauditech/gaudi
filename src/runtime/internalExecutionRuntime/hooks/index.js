/** Returns given value. Can be used for testing&debugging. */
module.exports.echo = function ({ value }) {
  return value;
};

/** Returns given value. Can be used for testing&debugging. */
module.exports.log = function ({ label, value }) {
  console.log(`[LOG]${label != null ? ` [${label}]:` : ":"} ${value}`);
};
