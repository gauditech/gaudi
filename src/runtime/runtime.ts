// import this file only with relative path because this file actually configures path aliases (eg @src, ...)
import "../common/setupAliases";

import { loadDefinition, readConfig } from "@src/runtime/config";
import { importHooks } from "@src/runtime/hooks";
import { createServer } from "@src/runtime/server/server";

// read environment
const config = readConfig();

const definition = loadDefinition(config.definitionPath);

importHooks(config);

// start server
createServer(definition, config);
