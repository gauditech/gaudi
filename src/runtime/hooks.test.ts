import { getInternalExecutionRuntimeName } from "@src/composer/executionRuntimes";
import { compileToOldSpec, compose } from "@src/index";
import { executeHook } from "@src/runtime/hooks";
import { CreateEndpointDef, Definition } from "@src/types/definition";

describe("hooks", () => {
  const def = createTestDefinition();

  /*
   * Using math division as a test operation:
   *  - tests multiple arguments
   *  - division is not cummutative operation and will detect errors if arguments are not passed in a specific order
   */

  describe("external hooks", () => {
    it("should resolve static value", async () => {
      const result = await executeHook(
        def,
        {
          runtimeName: "TestRuntime",
          code: { kind: "source", target: "divideStatic", file: "hooks.js" },
        },
        { x: 6, y: 2 }
      );

      expect(result).toBe(3);
    });

    it("should resolve promise value", async () => {
      const result = await executeHook(
        def,
        {
          runtimeName: "TestRuntime",
          code: { kind: "source", target: "divideAsync", file: "hooks.js" },
        },
        { x: 6, y: 2 }
      );

      expect(result).toBe(3);
    });
  });

  describe("inline hooks", () => {
    it("should resolve static value", async () => {
      const result = await executeHook(
        def,
        {
          runtimeName: "TestRuntime",
          code: { kind: "inline", inline: "x / y" },
        },
        { x: 6, y: 2 }
      );

      expect(result).toBe(3);
    });

    it("should resolve promise value", async () => {
      const result = await executeHook(
        def,
        {
          runtimeName: "TestRuntime",
          code: { kind: "inline", inline: "Promise.resolve(x / y)" },
        },
        { x: 6, y: 2 }
      );

      expect(result).toBe(3);
    });
  });

  describe("execution runtime", () => {
    it("should execute hooks from multiple exec runtimes", async () => {
      const bp = `
      runtime MathRuntime {
        default
        source path "./src/runtime/test/hooks"
      }

      runtime TextRuntime {
        source path "./src/runtime/test/hooks2"
      }

      model Result {
        field name { type string }
        field avg { type integer }
      }

      entrypoint Results {
        target Result
        create endpoint {
          action {
            create {
              set name hook {
                runtime TextRuntime
                arg value "First Last"
                arg prefix "Mr. "
                source prefix from "hooks2.js"
              }
              set avg hook {
                // thisi is from the default runtime
                arg x 100
                arg y 20
                source prefix from "hooks.js"
              }
            }
          }
        }
      }

    `;

      const result = compose(compileToOldSpec(bp));
      const action = result.entrypoints[0].endpoints
        .filter((ep): ep is CreateEndpointDef => ep.kind === "create")
        .shift()
        ?.actions.at(0);

      expect(action).toMatchSnapshot();
    });

    it("should run hook from internal exec runtime", async () => {
      const def = createTestDefinition();

      const result = await executeHook(
        def,
        {
          runtimeName: getInternalExecutionRuntimeName(),
          code: { kind: "source", file: "hooks/index.js", target: "echo" },
        },
        { value: "ASDF" }
      );

      expect(result).toBe("ASDF");
    });
  });
});

/**
 * Creates test definition struct
 */
function createTestDefinition(): Definition {
  const def = compose({
    entrypoints: [],
    models: [],
    populators: [],
    runtimes: [
      {
        name: "TestRuntime",
        default: true,
        sourcePath: "./src/runtime/test/hooks",
      },
    ],
    authenticator: undefined,
    generators: [],
  });

  return def;
}
