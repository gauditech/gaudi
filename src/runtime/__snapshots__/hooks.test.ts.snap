// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hooks execution runtime should execute hooks from multiple exec runtimes 1`] = `
{
  "alias": "$target_0",
  "changeset": [
    {
      "name": "name",
      "setter": {
        "args": [
          {
            "name": "value",
            "setter": {
              "kind": "literal",
              "type": "text",
              "value": "First Last",
            },
          },
          {
            "name": "prefix",
            "setter": {
              "kind": "literal",
              "type": "text",
              "value": "Mr. ",
            },
          },
        ],
        "hook": {
          "file": "hooks2.js",
          "kind": "source",
          "runtimeName": "TextRuntime",
          "target": "prefix",
        },
        "kind": "fieldset-hook",
      },
    },
    {
      "name": "avg",
      "setter": {
        "args": [
          {
            "name": "x",
            "setter": {
              "kind": "literal",
              "type": "integer",
              "value": 100,
            },
          },
          {
            "name": "y",
            "setter": {
              "kind": "literal",
              "type": "integer",
              "value": 20,
            },
          },
        ],
        "hook": {
          "file": "hooks.js",
          "kind": "source",
          "runtimeName": "MathRuntime",
          "target": "prefix",
        },
        "kind": "fieldset-hook",
      },
    },
  ],
  "kind": "create-one",
  "model": "Result",
  "select": [],
  "targetPath": [
    "$target_0",
  ],
}
`;
